### PR TITLE
fix(apps): harden Nostr bridge edge cases

### LINKED ISSUE #5334: fix(apps): harden remaining Nostr bridge NIP edge cases
## Context

PR #5318 added Android per-frame origin attestation for the Nostr app bridge and also fixed two pre-existing NIP-07 surface issues: removing the false `nip04` capability advertisement and allowing empty `signEvent.content`.

That PR comment also documented three remaining low-priority hardening items that were not fixed there. They are pre-existing bridge debt, not regressions from #5318, but they now need explicit tracking.

## Residual hardening items

1. **NIP-01 event serialization / C0 control characters**
   - Current behavior leans on Dart `json.encode`.
   - The review note says this differs from the strict NIP-01 "7 escapes only, rest verbatim" wording, while matching `nostr-tools` in practice.
   - Decide whether the bridge should keep this behavior with explicit conformance coverage or change serialization. If changed, prove event ID/signature compatibility.

2. **`nip44.decrypt` unsupported-version error shape**
   - Unsupported encrypted payload versions are currently flattened into a generic `decrypt_failed` result.
   - Preserve a more specific error if the signer/client boundary can expose one without leaking sensitive details or breaking existing callers.

3. **JSON `null` inside event tags**
   - A JSON `null` tag element is currently coerced to the string `"null"`.
   - Decide whether invalid tag elements should be rejected instead, or whether coercion is intentional and must be documented/tested.

## Expected outcome

The bridge should have explicit behavior and tests for each edge case above, rather than relying on implicit current behavior.

## Acceptance criteria

- Add focused tests in `mobile/packages/nostr_app_bridge_repository/test/nostr_app_bridge_service_test.dart` and/or sandbox bootstrap tests for the chosen behavior.
- If serialization changes, include a test that verifies event ID/signature compatibility for at least one event containing relevant control characters.
- If unsupported NIP-44 versions remain generic, document why at the service boundary and add a test locking that behavior.
- If `null` tag elements are rejected, verify the error returned to the app is stable and actionable; if they are coerced, document and test the coercion explicitly.
- Run the relevant bridge tests and `flutter analyze` from `mobile/` before closing.

## Source

Residual debt identified while auditing rabble-approved PRs from June 14-19, 2026. These items were documented in the #5318 PR discussion as low-priority hardening left for follow-up.

### PR DESCRIPTION
**Pull request by NotThatKindOfDrLiz**
## Summary

- normalize malformed bridge request payloads into stable invalid_request results at the bridge service boundary
- keep NIP-44 decrypt signer failures generic and non-leaky as decrypt_failed
- reject null event tag elements while preserving existing primitive tag coercion for compatibility
- pin a control-character event-id vector so current JSON-compatible serialization remains explicit

## Root cause

Issue #5334 tracked residual NIP edge cases left after #5318: implicit event serialization behavior, unsupported NIP-44 decrypt version error shape, and JSON null tag coercion. The bridge previously relied on caller-level exception handling for malformed request payloads and could leak signer exception text through the sandbox invalid_request path.

## Validation

- mise exec flutter@3.44.0 -- flutter analyze
- mise exec flutter@3.44.0 -- flutter test packages/nostr_app_bridge_repository/test/nostr_app_bridge_service_test.dart
- mise exec flutter@3.44.0 -- flutter test test/screens/apps/nostr_app_sandbox_screen_test.dart
- git push pre-push checks: merge conflict check, changed-file analyzer, changed bridge test

Closes #5334

### CODE REVIEW

**Review by rabble — APPROVED** [review]
**Approve.**

Security-positive hardening of the Nostr NIP-07 bridge: converts internal `ArgumentError`s into a controlled `_BridgeInvalidRequestException` caught at the dispatch boundary (so `invalid_request` carries only static field names, not exception text), prevents signer exception text leaking through `nip44.decrypt` (generic `decrypt_failed`, `errorMessage` null), rejects null tag elements, and pins control-char event-id serialization. Catches are narrow and intentional; tests assert real behavior; fixtures use full 64-char pubkeys (no Nostr ID truncation).

Nits:
- `[LOW]` `nostr_app_sandbox_screen.dart:484` still does `errorMessage: error.toString()` in the outer catch. Validation no longer reaches it after this PR, but any non-`Exception` `Error` (e.g. `StateError`, a JS-channel parse failure) thrown from `handleRequest` would still leak text to the web app — the stated goal isn't fully closed at that layer. Worth a follow-up.
- `[NIT]` A one-line comment on the pinned control-char id hash explaining it locks JSON-compatibility (not NIP-01 escaping correctness) would prevent future confusion.

File: mobile/lib/screens/apps/nostr_app_sandbox_screen.dart [hunk 1]
```diff
@@ -517,12 +517,23 @@ class _NostrAppSandboxScreenState extends ConsumerState<NostrAppSandboxScreen> {
...
         result: BridgeResult.error(
           'invalid_request',
-          errorMessage: error.toString(),
+          errorMessage: _safeBridgeErrorMessage(error),
         ),
       );
     }
   }
 
+  String? _safeBridgeErrorMessage(Object error) {
```
Comment (realmeylisdev) [inline mobile/lib/screens/apps/nostr_app_sandbox_screen.dart:535]: Nice approach sealing the leak at a single boundary with a static allowlist — this robustly drops the signer `Error`s (a malformed `nip44` pubkey throws `ArgumentError`/`RangeError`, which the service's `on Exception` doesn't catch) and the `jsonDecode` `FormatException`, while still surfacing the 3 helpful static bridge messages.

Minor: these 3 strings are now duplicated between the `throw FormatException(...)` sites (lines 477 / 499 / 502) and this allowlist. If one is ever reworded without the other, the message silently drops to `null` (the response stays correct, just less helpful). Extracting them into shared `const`s would keep the two in lockstep. Non-blocking.

File: mobile/packages/nostr_app_bridge_repository/lib/src/nostr_app_bridge_service.dart [hunk 2]
```diff
@@ -382,7 +398,14 @@ class NostrAppBridgeService {
       args['ciphertext'],
       fieldName: 'ciphertext',
     );
-    final plaintext = await signer.nip44Decrypt(pubkey, ciphertext);
+    final String? plaintext;
+    try {
+      plaintext = await signer.nip44Decrypt(pubkey, ciphertext);
+    } on Exception {
```
Comment (realmeylisdev) [inline mobile/packages/nostr_app_bridge_repository/lib/src/nostr_app_bridge_service.dart]: Defense-in-depth note for the package itself (the sandbox seal already covers the real app, so non-blocking): `on Exception` only catches `Exception`s, but the app-supplied `pubkey` flows into `NIP44V2.shareSecret` → `Kepler.byteSecret` *before* any payload parsing, and a malformed pubkey throws an `Error`, not an `Exception`. Verified empirically against the production signer:

- `'00'*32` → `ArgumentError: Invalid point compression`
- `'ff'*32` (x ≥ p) → `ArgumentError: Value x must be smaller than q`
- wrong/odd length → `ArgumentError: Incorrect length for compressed encoding`
- `…01` → `RangeError (end): Not in inclusive range 0..15: 63`

These escape this catch and `handleRequest` entirely (only `_BridgeInvalidRequestException` is caught), so a caller without the sandbox seal would see a generic `invalid_request` rather than `decrypt_failed`. `_handleNip44Encrypt` (line 382) and `_handleGetRelays` (line 308) have no guard at all and behave the same. If you want the bridge package to be self-contained regardless of caller, broadening these to `catch (_)` → `decrypt_failed`/`encrypt_failed` (or validating the pubkey is 64-char hex up front) would do it.

File: mobile/packages/nostr_app_bridge_repository/lib/src/nostr_app_bridge_service.dart [hunk 3]
```diff
@@ -246,17 +255,24 @@ class NostrAppBridgeService {
...
+        case 'getRelays':
+          result = await _handleGetRelays();
+        case 'signEvent':
+          result = await _handleSignEvent(args);
+        case 'nip44.encrypt':
+          result = await _handleNip44Encrypt(args);
+        case 'nip44.decrypt':
+          result = await _handleNip44Decrypt(args);
+      }
+    } on _BridgeInvalidRequestException catch (error) {
```
Comment (realmeylisdev) [inline mobile/packages/nostr_app_bridge_repository/lib/src/nostr_app_bridge_service.dart:271]: Catching the typed exception here is a real improvement — cleaner message (no `ArgumentError`'s `Invalid argument(s):` prefix) and these failures now get an audit row (previously the `ArgumentError` propagated past `_recordAudit` and wasn't recorded).

One thing to confirm is intent: when this catch fires, `auditDecision` is still `allowed`/`promptAllowed`, so a malformed request is audited as an *allowed* decision with `errorCode: 'invalid_request'`. That's a defensible reading (the capability was allowed; the payload was just malformed), and for `signEvent` the user is also prompted before this validation runs — just flagging the new audit semantics. Non-blocking.

File: mobile/packages/nostr_app_bridge_repository/lib/src/nostr_app_bridge_service.dart [hunk 4]
```diff
@@ -444,17 +471,28 @@ class NostrAppBridgeService {
...
           }
-          return tag.map((item) => item.toString()).toList(growable: false);
+          return tag
+              .map((item) {
+                if (item == null) {
+                  throw const _BridgeInvalidRequestException(
+                    'event.tags must not contain null elements',
+                  );
+                }
+                return item.toString();
```
Comment (realmeylisdev) [inline mobile/packages/nostr_app_bridge_repository/lib/src/nostr_app_bridge_service.dart]: Rejecting `null` here is good. The remaining `item.toString()` still coerces any other non-string element, and `toString()` on some JSON types is surprising: a JSON number `42.0` → `'42.0'` (note the `.0`), a nested array `['x']` → `'[x]'`, an object `{'a':1}` → `'{a: 1}'`. The new "keeps primitive tag coercion" test only exercises `int` (`42`) and `bool` (`true`).

If the intent is "strings + safely-coercible scalars", consider rejecting non-`(String|num|bool)` elements (and deciding how an int-valued double should render); otherwise a short doc/test note on the coercion would help. Nit.

**Review by realmeylisdev — APPROVED** [review]
## Formal review — #5339

Reviewed the full cumulative diff (both commits) plus the surrounding bridge / signer / sandbox code, ran the bridge test suite locally (12/12 pass), traced the signer crypto path empirically, and verified the cross-repo event-id interop against `divine-funnelcake`, `divine-web`, `divine-signer`, and `keycast`. All CI green on the head (incl. Tests).

**LGTM — approving.** This is a careful, well-tested hardening pass, and the follow-up `seal sandbox bridge error boundary` commit cleanly closes the one substantive concern: `_handleBridgeMessage` no longer echoes `error.toString()` — `_safeBridgeErrorMessage` passes through only an allowlist of 3 static bridge messages and drops everything else (signer `Error`s, `jsonDecode` `FormatException`s), with a sandbox test pinning it. The comments below are nits + one cross-repo heads-up, none blocking.

### Verified correct
- **Event-id serialization (issue item 1).** `nostr_sdk/event.dart:183` computes the id via `json.encode([0,…])` → `utf8.encode` → `sha256`, escaping control chars as `\uXXXX`. This matches the dominant ecosystem — nostr-tools (divine-web), rust-nostr (keycast), serde_json — so "keep `json.encode` + pin it" is the interoperable choice. The pinned id `bf207992…` is reproducible (test passes).
- **Error boundary.** The `_read*` validation messages are all static field names (no user data) and are delivered via the service *result* (not the sealed catch), so they survive; only unexpected throws are sealed. Confirmed the seal drops `ArgumentError`/`RangeError`/`StateError`/non-allowlisted `FormatException` and omits the `"message"` key entirely.
- **Scope.** Confined to the active `nostr_app_bridge_repository`; `packages/nostr_apps` is dead/legacy and untouched. No caller relied on the old `ArgumentError`.

### Non-blocking (see inline)
1. Seal: the 3 safe strings are duplicated between the `throw FormatException(...)` sites and the allowlist — extracting shared `const`s keeps them in lockstep.
2. Defense-in-depth: the bridge *package* still lets signer `Error`s (malformed pubkey) escape `handleRequest`, so used outside the sealed sandbox they'd surface as generic `invalid_request` rather than `decrypt_failed`.
3. Audit decision for a caught `invalid_request` is recorded as `allowed`/`promptAllowed` — confirm intended.
4. Tag coercion still stringifies doubles / nested values via `toString()`; the new test covers only int + bool.

### Cross-repo heads-up (informational — not this PR, mobile side is correct)
While verifying interop I found a real divergence on divine's own relay: `divine-funnelcake/crates/proto/src/relay_event.rs:99-109` (`RelayEvent::verify`, invoked on ingest at `crates/relay/src/relay.rs:1673`, rejecting on `InvalidId`) recomputes the content with a hand-rolled serializer that escapes **only** `\`, `"`, `\n` and leaves `\t`, `\r`, `\b`, `\f` and all other control chars verbatim. So an event whose content contains a tab or control char — exactly what this PR's test pins — would be recomputed to a different id and **rejected by the funnelcake relay**. That's a pre-existing funnelcake-side NIP-01 under-escaping bug affecting all divine clients (the mobile side here is the spec-and-ecosystem-correct one), so it's out of scope for this PR — but it means "JSON-compatible event id" doesn't, today, imply divine-relay acceptance for control-char content. Worth a separate funnelcake issue to align its serializer with `serde_json`.

### Acceptance criteria (#5334)
- Item 1 (serialization conformance): covered by the id-pinning test (optional: also assert the signature / round-trip).
- Item 2 (NIP-44 unsupported version stays generic): covered + documented; the malformed-pubkey variant is sealed at the boundary (inline #2).
- Item 3 (null tag elements rejected): covered (reject) + coercion documented by test.
